### ISSUE TITLE
Request for docs: How to runUpdate with records?

### ISSUE THREAD

**Issue by saurabhnanda** [tomjaguarpaw/haskell-opaleye#205]
I'm not able to figure out how to do a `runUpdate` given the following record-types:

```
data TenantPoly key createdAt updatedAt name status ownerId backofficeDomain = Tenant {
  tenantKey :: key
  ,tenantCreatedAt :: createdAt
  ,tenantUpdatedAt :: updatedAt
  ,tenantName :: name
  ,tenantStatus :: status
  ,tenantOwnerId :: ownerId
  ,tenantBackofficeDomain :: backofficeDomain
  }

type TenantPGWrite = TenantPoly
  (Maybe (Column PGInt8)) -- key
  (Maybe (Column PGTimestamptz)) -- createdAt
  (Maybe (Column PGTimestamptz)) -- updatedAt
  (Column PGText) -- name
  (Column PGText) -- status
  (Column (Nullable PGInt8)) -- ownerId
  (Column PGText) -- backofficeDomain

type TenantPGRead = TenantPoly
  (Column PGInt8) -- key
  (Column PGTimestamptz) -- createdAt
  (Column PGTimestamptz) -- updatedAt
  (Column PGText) -- name
  (Column PGText) -- status
  (Column (Nullable PGInt8)) -- ownerId
  (Column PGText) -- backofficeDomain
```

**How do I make the following function type-check?**

```
activateTenant :: TenantId -> AppM (Tenant)
activateTenant tenantId@(TenantId tid) = do
  conn <- askDbConnection
  liftIO $ do 
    _ <- runUpdate conn tenantTable
      (\tenant -> tenant & status .~ TenantActive) -- ERROR: Couldn't match type ‘Column PGInt8’ with ‘Maybe (Column PGInt8)’
      (\tenant -> (tenant ^. key .== pgInt8 tid))
    head <$> runQuery conn (tenantById tenantId)
```

If I understand the type error correctly, it's expecting that lambda to return a `TenantPGWrite`, where `key`, `createdAt`, `updatedAt` are Maybes. Does that mean I need to transform the entire row from `TenantPGRead` to `TenantPGWrite` for a simple update?

Also, it's not clear from the type signature and [docs for `runUpdate`](https://www.stackage.org/haddock/lts-6.12/opaleye-0.4.2.0/Opaleye-Manipulation.html#v:runUpdate) what the resulting `Int64` represents. Is it the number of records updated?

**Comment by saurabhnanda** [comment]
Any help with this? Should I clarify my question further?

**Comment by tomjaguarpaw** [comment]
> Does that mean I need to transform the entire row from TenantPGRead to TenantPGWrite for a simple update?

Yes

> what the resulting Int64 represents. Is it the number of records updated?

Yes

**Comment by tomjaguarpaw** [comment]
>    what the resulting Int64 represents. Is it the number of records updated?

By the way, I've fixed the docs in the latest version.  Thanks for the report!


https://hackage.haskell.org/package/opaleye-0.5.1.1/docs/Opaleye-Manipulation.html

**Comment by saurabhnanda** [comment]
>> Does that mean I need to transform the entire row from TenantPGRead to TenantPGWrite for a simple update?

> Yes

Is there any way to NOT have to do this? Isn't it tedious?

Also, wouldn't Opaleye already have a mechanism to convert a `TenantPGRead` to a `TenantPGWrite`, as an offshoot of the set-up/mapping process?

**Comment by tomjaguarpaw** [comment]
I'm sure there exists a nice convenient solution waiting to be implemented, but regrettably I have many competing demands on my time.

**Comment by saurabhnanda** [comment]
> I'm sure there exists a nice convenient solution waiting to be implemented, but regrettably I have many competing demands on my time.

Completely understand and appreciate that @tomjaguarpaw As the primary author of the library, if you agree that the update API could be better/easier to use, can you pen down your thoughts about how to make it better? It can guide other contributors in the direction fitting the overall design vision for the library.

If I find the time after my current quest, even I might take it up.

**Comment by tomjaguarpaw** [comment]
Yes, I'm very happy to write down *how* to make improvements (although that also requires time) but *what* to improve is up to you :)

**Comment by tomjaguarpaw** [comment]
By the way, this particular piece of functionality requires knowing nothing about Opaleye at all.  You just want some polymorphic way of converting between similar datatypes, probably based on a typeclass. You'll probably find `Data.Profunctor.Product.Default` useful.

**Comment by saurabhnanda** [comment]
Is there a typesafe version of `UPDATE` that deals with only specific columns (analogous to typesafe `SELECT`s)?

**Comment by tomjaguarpaw** [comment]
I'm not sure what you mean.  What are you finding not typesafe about it?

**Comment by saurabhnanda** [comment]
It's not about **typesafe** per-se, it's about dealing with only **specific columns** and not the entire row. Else, one has to resort to raw SQL, which is not typesafe.

**Comment by tomjaguarpaw** [comment]
I'm still not sure what you mean.  `runUpdate` will happily change only specific columns.

**Comment by bergmark** [comment]
I don't think this is possible, see #92. You have to give values to every column where a default isn't declared

**Comment by tomjaguarpaw** [comment]
Right, so you just give it the same value it started with.

````haskell
runUpdate table (\(id_, x, y, s) -> (id_, x + y, x - y, s))
                                  (\(id_, _, _, _) -> id_ .== 5)
````

will leave the `id_` column as it is.

**Comment by saurabhnanda** [comment]
#92 is a long and interesting read. I think I'm facing the exact same problem. I don't want anything to do with columns that my `UPDATE` is not touching.

**Comment by saurabhnanda** [comment]
I think we can close this issue and take the party over to #92

**Comment by tomjaguarpaw** [comment]
If you like, but I think that party's over!

**Comment by saurabhnanda** [comment]
Isn't that issue still open? IMO it's a very important part of the API and **should** be fixed. I'll think about the problem and post there, if I can add something meaningful to that discussion.

**Comment by tomjaguarpaw** [comment]
IMHO it is fixed :) but sure, feel free to go to #92 and discuss more.

**Comment by tomjaguarpaw** [comment]
If you need more help with this please reopen.  Probably best to discuss further on #92.
